com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/linking/LinkingService.java

```diff
@@ -64,70 +65,114 @@
+    if (exports.hasNext()) {
+      QualifiedName targetName = exports.next().getName();
+      if (targetName != null && !targetName.isEmpty()) {
+        ImportedNamesTypesAdapter adapter = getImportedNamesAdapter(context);
+        final QualifiedName lowerCaseName = targetName.toLowerCase();// NOPMD targetName not a String!
+        adapter.getImportedNames().add(lowerCaseName);
+        if (adapter.getImportedNamesTypes().containsKey(lowerCaseName)) {
+          adapter.getImportedNamesTypes().get(lowerCaseName).add(type);
+        } else {
+          adapter.getImportedNamesTypes().put(lowerCaseName, Sets.newHashSet(type));
         }
       }
-    } else if (name != null && name.length() > 0) { // import parsed string for unresolved references
-      QualifiedName unresolvedName = crossRefHelper.toUnresolvedReferenceName(name);
-      adapter.getImportedNames().add(unresolvedName);
-      if (adapter.getImportedNamesTypes().containsKey(unresolvedName)) {
-        adapter.getImportedNamesTypes().get(unresolvedName).add(type);
+    }
+  }
+
+  /**
+   * Register a name as imported.
+   *
+   * @param context
+   *          context object within which a reference is being set to some object, must not be {@code null}
+   * @param desc
+   *          the associated description. Its name is recorded as imported, must not be {@code null}
+   * @param type
+   *          the lookup type, may be {@code null}
+   */
+  public void importObject(final EObject context, final IEObjectDescription desc, final EClass type) {
+    QualifiedName targetName;
+    if (desc instanceof AliasingEObjectDescription) {
+      targetName = ((AliasingEObjectDescription) desc).getOriginalName();
+    } else {
+      targetName = desc.getName();
+    }
+    if (targetName != null && !targetName.isEmpty()) {
+      ImportedNamesTypesAdapter adapter = getImportedNamesAdapter(context);
+      final QualifiedName lowerCaseName = targetName.toLowerCase();// NOPMD targetName not a String!
+      adapter.getImportedNames().add(lowerCaseName);
+      if (adapter.getImportedNamesTypes().containsKey(lowerCaseName)) {
+        adapter.getImportedNamesTypes().get(lowerCaseName).add(type);
       } else {
-        adapter.getImportedNamesTypes().put(unresolvedName, Sets.newHashSet(type));
+        adapter.getImportedNamesTypes().put(lowerCaseName, Sets.newHashSet(type));
       }
     }
   }
 
+  /**
+   * Register a typed unresolved reference.
+   *
+   * @param context
+   *          context object within which a reference is being set to unresolved, must not be {@code null}
+   * @param name
+   *          the lookup name, may be {@code null}
+   * @param type
+   *          the lookup type, may be {@code null}
+   */
+  public void registerUnresolvedReference(final EObject context, final String name, final EClass type) {
+    ImportedNamesTypesAdapter adapter = getImportedNamesAdapter(context);

```

Comment:
i guess we can factor this code out in a separate method... (everything except the 2nd line)

Reply:
Done.